In [1]:
import pandas as pd
df = pd.read_csv('lab7.csv', sep='\t', header=None)

In [2]:
# Выведем весь датасет и проанализируем имеющийся датасет
df

0
0    228798439247474688 2012-07-27 14:26:13 MSK <ro...
1    229908598749270016 2012-07-30 15:57:36 MSK <ro...
2    230251866020253697 2012-07-31 14:41:37 MSK <ro...
3    230660623401377792 2012-08-01 17:45:52 MSK <ro...
4    231015933769043969 2012-08-02 17:17:45 MSK <ro...
..                                                 ...
492  364647925806141440 2013-08-06 11:23:35 MSK <ro...
493  364690187684544512 2013-08-06 14:11:31 MSK <ro...
494  364993573852151808 2013-08-07 10:17:03 MSK <ro...
495  365012870121005056 2013-08-07 11:33:44 MSK <ro...
496  365080585284235264 2013-08-07 16:02:49 MSK <ro...

[497 rows x 1 columns]

Имеющийся датасет содержит 1 атрибут и 497 строк

### Откройте файл и присвойте названия полученным столбцам, используя параметр df.columns. 

In [3]:
# Переименуем имеющийся атрибут
df.columns = ['Information']
df

Information
0    228798439247474688 2012-07-27 14:26:13 MSK <ro...
1    229908598749270016 2012-07-30 15:57:36 MSK <ro...
2    230251866020253697 2012-07-31 14:41:37 MSK <ro...
3    230660623401377792 2012-08-01 17:45:52 MSK <ro...
4    231015933769043969 2012-08-02 17:17:45 MSK <ro...
..                                                 ...
492  364647925806141440 2013-08-06 11:23:35 MSK <ro...
493  364690187684544512 2013-08-06 14:11:31 MSK <ro...
494  364993573852151808 2013-08-07 10:17:03 MSK <ro...
495  365012870121005056 2013-08-07 11:33:44 MSK <ro...
496  365080585284235264 2013-08-07 16:02:49 MSK <ro...

[497 rows x 1 columns]

Атрибут был успешно переименован и теперь называется 'Information'

In [4]:
# Выведем первую строку атрибута Information
df.Information[0]

'228798439247474688 2012-07-27 14:26:13 MSK <roscomnadzor> Все карты покрытия дорог России мобильной связью на сайте Роскомнадзора http://t.co/VE4StA0e'

Необходимо избавиться от лишней информации в строках так, чтобы остался только текст типа: "Все карты покрытия дорог России мобильной связью на сайте Роскомнадзора" 

### Произведите токенизацию текста для каждой записи

In [5]:
import re
import string
from nltk.tokenize import word_tokenize 

In [6]:
# Произведем токенизацию исходного набора для каждой записи. 
# Для разделения текста на отдельные слова и знаки воспользуемся метод word_tokenize().
# Создадим функцию clean_text при помощи инструкции def и построчно заменим символы, 
# от которых необходимо избавиться, на пустые значения.
def clean_text(tweet):
    tweet.lower()
    tweet = re.sub(r"http\S+|www\S+|http\S+", '', tweet, flags = re.MULTILINE)
    tweet = re.sub(r"\@\w+|\#", '', tweet)
    tweet = re.sub(r"MSK", '', tweet)
    tweet = re.sub(r"roscomnadzor", '', tweet)
    tweet = re.sub(r"[0-9]", '', tweet)
    tweet = tweet.translate(str.maketrans('', '', string.punctuation))

    tweet_tokens = word_tokenize(tweet)
    filtred_words = [w for w in tweet_tokens]
    return ' '.join(filtred_words)

In [7]:
# Применяем функцию clean_text для нашего атрибута в датасете
df.Information = df.Information.apply(clean_text)

In [8]:
# Повторно выведем первую строку атрибута Information
df.Information[0]

'Все карты покрытия дорог России мобильной связью на сайте Роскомнадзора'

По полученному результату можно сделать вывод, что токенизация исходного набора данных для каждой записи произведена успешно.

In [9]:
#from nltk import pos_tag
# nltk.download()

### Проведите стемминг исходных слов и измерьте время выполнения операции. Сохраните результат стемминга в переменную. 

In [10]:
# Создаем новую переменную stem_df и дадим название его единственного атрибута 'tweets'
stem_df = pd.DataFrame(columns = ['tweets'])
stem_df

Empty DataFrame
Columns: [tweets]
Index: []

In [11]:
from nltk.stem.snowball import SnowballStemmer
from datetime import timedelta
import time

In [12]:
# Создадим функцию stemming_text при помощи инструкции def.
# Лемматизация использует словарь и морфологический анализ, 
# чтобы в итоге привести слово к его канонической форме – лемме.  
def stemming_text(tweet):
    tweet_tokens = word_tokenize(tweet)
    filtred_words = [w for w in tweet_tokens]
    snow_lem = SnowballStemmer('russian')
    stemmed_words = [snow_lem.stem(w) for w in filtred_words]
    return ' '.join(stemmed_words)

In [13]:
# Измерим время выполнения операции.
start_time = time.monotonic()
stem_df.tweets = df.Information.apply(stemming_text)
end_time = time.monotonic()
print(timedelta(seconds = end_time - start_time))

0:00:00.593000


In [14]:
stem_df.tweets[0]

'все карт покрыт дорог росс мобильн связ на сайт роскомнадзор'

По полученному результату можно сделать вывод, что стемминг исходных слов набора данных произведен успешно.
В ходе стемминга происходит отсечение частей слова, таким образом, что от слова остается только корень, что и можно наблюдать в ячейке [28].

### Проведите лемматизацию исходных слов и измерьте время выполнения операции. Сравните время выполнения стемминга и лемматизации. 

Лемматизация – процесс, который использует словарь и морфологический анализ, чтобы в итоге привести слово к его канонической форме – лемме. Для лемматизации русских и украинских текстов используется библиотека pymorphy2. 
Метод normal_forms() из класса MorphAnalyzer() позволяет нормализовать слово, приведенное в параметрах.

In [15]:
!pip install pymorphy2

In [16]:
lemmed_df = pd.DataFrame(columns = ['lem_tweets'])
lemmed_df

Empty DataFrame
Columns: [lem_tweets]
Index: []

In [17]:
import pymorphy2

def lemming_text(tweet):
    tweet_tokens = word_tokenize(tweet)
    filtred_words = [w for w in tweet_tokens]
    lemmatizer = pymorphy2.MorphAnalyzer()
    lemmed_words = [lemmatizer.normal_forms(w)[0] for w in filtred_words]
    return ' '.join(lemmed_words)

In [18]:
# Измерим время выполнения операции.
start_time = time.monotonic()
lemmed_df.lem_tweets = df.Information.apply(lemming_text)
end_time = time.monotonic()
print(timedelta(seconds = end_time - start_time))

0:01:59.297000


In [19]:
lemmed_df.lem_tweets[0]

'всё карта покрытие дорога россия мобильный связь на сайт роскомнадзор'

Процесс выполенния операции леммитизации занимает примерно в два раза больше времени.

### Создайте словари из 50, 500 и 1000 слов по результатам обработки с использованием стемминга и лемматизации. Сравните словари с одинаковым количеством слов. Какой метод лучше работает для приведенного набора данных?

Словарь представляет собой набор самых популярных слов датасета. При этом значением преобразованной строки будет массив чисел, обозначающий частоту вхождения слова из словаря в данный текст.
Для формирования словаря используется метод fit_transform(), который имеет следующие параметры.
Для вывода слов, вошедших в словарь, используется метод get_feature_names().
Для получения массива чисел необходимо преобразовать результат выполнения метода fit_transform() в массив, для чего используется функция to_array().

In [20]:
from sklearn.feature_extraction.text import CountVectorizer

In [21]:
# Создадим словарь из 500 слов по результатам обработки с использованием стемминга
from sklearn.feature_extraction.text import CountVectorizer
cv500 = CountVectorizer(analyzer = 'word', ngram_range = (1,1), min_df = 1, max_features = 500)
X_transform500 = cv500.fit_transform(stem_df.tweets)
print(cv500.get_feature_names())

['ip', 'ngn', 'online', 'rt', 'vpikovru', 'youtube', 'zapretinforu', 'абонент', 'аварийн', 'автодорог', 'автоматизирова', 'автор', 'администрац', 'адрес', 'ажар', 'аккредитац', 'александр', 'анонс', 'апрел', 'базов', 'бегущ', 'без', 'безопасн', 'билайн', 'ближайш', 'блокировк', 'бол', 'буд', 'будет', 'будут', 'был', 'быт', 'бюджетн', 'важн', 'вам', 'ваш', 'ввишневецк', 'ввоз', 'веден', 'верс', 'вещан', 'вещател', 'взаимодейств', 'вид', 'видеозап', 'википед', 'включ', 'внес', 'внесен', 'внесл', 'вниман', 'внутрен', 'во', 'возможн', 'возобнов', 'вопрос', 'врем', 'времен', 'все', 'всег', 'встреч', 'встреча', 'втор', 'вчер', 'вчу', 'вы', 'выда', 'вынес', 'газет', 'глав', 'го', 'говор', 'год', 'гражда', 'дан', 'два', 'двух', 'действ', 'дел', 'дет', 'детск', 'деятельн', 'диалог', 'для', 'дни', 'до', 'договор', 'доклад', 'документ', 'долж', 'домен', 'доступ', 'друг', 'еаис', 'есл', 'ест', 'ещ', 'жалоб', 'жар', 'за', 'заверш', 'завтр', 'заключен', 'закон', 'законодательств', 'закрыт', 'замечан

In [22]:
# Создадим словарь из 50 слов по результатам обработки с использованием стемминга
cv50 = CountVectorizer(analyzer = 'word', ngram_range = (1,1), min_df = 1, max_features = 50)
X_transform50 = cv50.fit_transform(stem_df.tweets)
print(cv50.get_feature_names())

['аварийн', 'адрес', 'вещан', 'вопрос', 'вчу', 'выда', 'вынес', 'дан', 'действ', 'дет', 'для', 'за', 'закон', 'запис', 'запрещен', 'зарегистрирова', 'заседан', 'защ', 'информац', 'конкурс', 'лиценз', 'минувш', 'на', 'не', 'недел', 'нов', 'област', 'оператор', 'персональн', 'по', 'предписан', 'предупрежден', 'приостанов', 'прошл', 'регистрац', 'реестр', 'решен', 'ркн', 'роскомнадзор', 'росс', 'рэс', 'сайт', 'свидетельств', 'связ', 'сми', 'совет', 'телерадиовещан', 'телерадиовещател', 'экспертн', 'эт']


In [23]:
# Создадим словарь из 1000 слов по результатам обработки с использованием стемминга
cv500 = CountVectorizer(analyzer = 'word', ngram_range = (1,1), min_df = 1, max_features = 1000)
X_transform500 = cv500.fit_transform(stem_df.tweets)
print(cv500.get_feature_names())

['best', 'common', 'cообщ', 'durixru', 'fzru', 'golixru', 'google', 'ii', 'ip', 'ipадрес', 'iv', 'lte', 'neslandru', 'netgeo', 'newslandru', 'ngn', 'ofcom', 'online', 'practice', 'rsocinru', 'rt', 'ru', 'snob', 'twitter', 'uk', 'vpikovru', 'youtube', 'zapretinforu', 'абонент', 'авансов', 'авар', 'аварийн', 'август', 'автодорог', 'автоматизирова', 'автономн', 'автор', 'админ', 'администратор', 'администрац', 'адрес', 'ажар', 'ажаров', 'акад', 'аккаунт', 'аккредитац', 'акт', 'активирова', 'активн', 'актуализирова', 'александр', 'ан', 'анализ', 'анонс', 'антимонопольн', 'антипиратск', 'апологет', 'апрел', 'атак', 'аудитор', 'баз', 'базов', 'бегущ', 'без', 'безопасн', 'безусловн', 'беспилотн', 'билайн', 'благодар', 'благополуч', 'бланк', 'ближайш', 'блокировк', 'бол', 'больш', 'бс', 'буд', 'будет', 'будут', 'бы', 'был', 'быт', 'бюджет', 'бюджетн', 'важн', 'вам', 'вас', 'ваш', 'введ', 'введен', 'ввишневецк', 'ввоз', 'веден', 'велик', 'великобритан', 'верн', 'верс', 'ветеран', 'вещан', 'веща

In [24]:
# Создадим словарь из 500 слов по результатам обработки с использованием лемминга
from sklearn.feature_extraction.text import CountVectorizer
cv500 = CountVectorizer(analyzer = 'word', ngram_range = (1,1), min_df = 1, max_features = 500)
X_transform500 = cv500.fit_transform(lemmed_df.lem_tweets)
print(cv500.get_feature_names())

['ip', 'ngn', 'online', 'rt', 'vpikovru', 'youtube', 'zapretinforu', 'абонент', 'аварийный', 'авария', 'автодорога', 'автоматизированный', 'автор', 'администрация', 'адрес', 'ажар', 'аккредитация', 'александр', 'анонс', 'апрель', 'базовый', 'бежать', 'без', 'безопасность', 'билайн', 'близкий', 'блокировка', 'более', 'быть', 'бюджетный', 'важно', 'ваш', 'ввишневецкий', 'ввоз', 'ведение', 'версия', 'весь', 'вещание', 'вещатель', 'взаимодействие', 'вид', 'видеозапись', 'видеть', 'википедия', 'включить', 'внесение', 'внести', 'внимание', 'внутренний', 'возможно', 'возобновить', 'вопрос', 'воспользоваться', 'временной', 'время', 'встреча', 'встречаться', 'всё', 'второй', 'вчера', 'вчу', 'вы', 'выдать', 'выдача', 'вынести', 'газета', 'глава', 'го', 'говорить', 'год', 'гражданин', 'грчца', 'да', 'давать', 'далее', 'данные', 'дать', 'два', 'действие', 'действовать', 'дело', 'день', 'детский', 'деятельность', 'диалог', 'для', 'до', 'договор', 'доклад', 'документ', 'должный', 'доменный', 'доступ

In [25]:
# Создадим словарь из 50 слов по результатам обработки с использованием лемминга
from sklearn.feature_extraction.text import CountVectorizer
cv500 = CountVectorizer(analyzer = 'word', ngram_range = (1,1), min_df = 1, max_features = 50)
X_transform500 = cv500.fit_transform(lemmed_df.lem_tweets)
print(cv500.get_feature_names())

['аварийный', 'адрес', 'быть', 'вещание', 'вопрос', 'вчу', 'выдать', 'вынести', 'действие', 'для', 'за', 'закон', 'запись', 'запретить', 'зарегистрировать', 'заседание', 'защита', 'информация', 'конкурс', 'лицензия', 'мы', 'на', 'не', 'неделя', 'новый', 'область', 'один', 'оператор', 'персональный', 'по', 'предписание', 'предупреждение', 'приостановить', 'прошлый', 'ребёнок', 'регистрация', 'реестр', 'решение', 'ркн', 'роскомнадзор', 'россия', 'рэс', 'сайт', 'свидетельство', 'связь', 'сми', 'совет', 'телерадиовещание', 'телерадиовещатель', 'экспертный']


In [26]:
# Создадим словарь из 1000 слов по результатам обработки с использованием лемминга
from sklearn.feature_extraction.text import CountVectorizer
cv500 = CountVectorizer(analyzer = 'word', ngram_range = (1,1), min_df = 1, max_features = 1000)
X_transform500 = cv500.fit_transform(lemmed_df.lem_tweets)
print(cv500.get_feature_names())

['best', 'common', 'cообщит', 'durixru', 'fzru', 'golixru', 'google', 'ii', 'ip', 'ipадрес', 'iv', 'lte', 'neslandru', 'netgeo', 'newslandru', 'ngn', 'nвопрос', 'ofcom', 'online', 'practice', 'rsocinru', 'rt', 'ru', 'snob', 'twitter', 'uk', 'vpikovru', 'vpikovrunый', 'youtube', 'zapretinforu', 'абонент', 'авансовый', 'аварийный', 'авария', 'август', 'автодорога', 'автоматизированный', 'автономный', 'автор', 'админ', 'администратор', 'администрация', 'адрес', 'ажар', 'ажаров', 'акадо', 'аккаунт', 'аккредитация', 'акт', 'активировать', 'активный', 'актуализировать', 'александр', 'анализ', 'ано', 'анонс', 'антимонопольный', 'антипиратский', 'апологет', 'апрель', 'атака', 'аудитория', 'база', 'базовый', 'бежать', 'без', 'безопасность', 'безопасный', 'безусловно', 'беспилотный', 'билайн', 'благодарить', 'благополучие', 'бланк', 'близкий', 'блокировка', 'более', 'большой', 'бс', 'бы', 'быть', 'бюджет', 'бюджетный', 'важно', 'ваш', 'введение', 'ввести', 'ввишневецкий', 'ввоз', 'ведение', 'вел

Вывод: оба метода при создании словарей с одинаковым количеством слов показывают примерно одинаковый результат.